### DISCUSSION TITLE
How to get the output directory from inside a plugin?

### DISCUSSION THREAD

**Discussion by rezaalavi** [11ty/eleventy#2483] [Q&A]
I am writing a plugin that generates files that need to be written to the output directory.
I can't find any method to get the full path to the output directory. Specifically, if the path is passed with --output argument to elements.

Can someone please point me in the right direction?

Thanks.

**Comment by pdehaan** [comment]
I found a `dir` in the `eleventyConfig` object from within my fake plugin. But that only seems to return whatever I set in my root .eleventy.js config file, and doesn't account for <kbd>--output</kbd> being set via the command line.

I also found https://github.com/11ty/eleventy/issues/2320#issuecomment-1097291160 from @zachleat, which says:
> I think there is also a `eleventy.directories` event that isn’t on the docs but you shouldn’t need that when `eleventy.before` has the same info.

I'm testing in @11ty/eleventy@canary (v2.0.0-canary.12) per https://www.11ty.dev/docs/events/#event-arguments "Coming soon in v2.0.0" comment; but am seeing subtle differences between the [undocumented] `eleventy.directories` and `eleventy.before` results:

```sh
> eleventy --output=bananasinpajamas

plugin: { input: 'src', includes: '_includes', data: '_data', output: 'www' }
{
  'eleventy.directories': {
    input: 'src',
    data: 'src/_data',
    includes: 'src/_includes',
    layouts: undefined,
    output: 'bananasinpajamas'
  }
}
{
  'eleventy.before': {
    inputDir: 'src',
    dir: {
      input: 'src',
      includes: '_includes',
      data: '_data',
      output: 'www'
    },
    runMode: 'build',
    outputMode: 'fs'
  }
}
[11ty] Writing bananasinpajamas/index.html from ./src/index.njk
[11ty] Wrote 1 file in 0.05 seconds (v2.0.0-canary.12)
```

Where my custom plugin is:
```js
module.exports = function (eleventyConfig, pluginConfig={}) {
  console.log(`plugin: ${inspect(eleventyConfig.dir)}`);

  eleventyConfig.on('eleventy.before', async (data) => {
    console.log({"eleventy.before": data});
  });

  eleventyConfig.on('eleventy.directories', async (data) => {
    console.log({"eleventy.directories": data});
  });
};
```

and my .eleventy.js looks like this:
```js
module.exports = function (eleventyConfig) {
  eleventyConfig.addPlugin(require("./plugin"));
  eleventyConfig.addFilter("inspect", require("node:util").inspect);

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

So, you can see that my .eleventy.js config file sets the `{input: "src", output: "www"}` directories, which I then override the output directory by calling <kbd>eleventy --output=bananasinpajamas</kbd>.
The `eleventy.directories` event correctly sets my `output` directory to 'bananasinpajamas'.
But the `eleventy.before` event seems to still use the `output` value of "www" from the .eleventy.js config file, ignoring the <kbd>--output</kbd> directory from the command line.

I still can't seem to find an easier way to get the fully resolved output directory from within a plugin, since using the `eleventy.directories` event might not be optimal.

**Reply by pdehaan** [reply]
I think I hate myself, but here's some possible crimes-against-code which converts the EventEmitter into a Promise, so you can `await` the `eleventy.directories` results in your plugin:

```js
module.exports = async function (eleventyConfig, pluginConfig={}) {
  const dirs = await emitterToPromise(eleventyConfig, "eleventy.directories");
  console.log(dirs);
};

async function emitterToPromise(cfg, event) {
  return new Promise((resolve) => cfg.on(event, resolve));
}
```

```sh
> eleventy

{
  input: 'src',
  data: 'src/_data',
  includes: 'src/_includes',
  layouts: undefined,
  output: 'www'
}
[11ty] Writing www/index.html from ./src/index.njk
[11ty] Wrote 1 file in 0.06 seconds (v2.0.0-canary.12)
```
&hellip; and with an overridden <kbd>--output</kbd> directory:
```sh
> eleventy --output=bananasinpajamas

{
  input: 'src',
  data: 'src/_data',
  includes: 'src/_includes',
  layouts: undefined,
  output: 'bananasinpajamas'
}
[11ty] Writing bananasinpajamas/index.html from ./src/index.njk
[11ty] Wrote 1 file in 0.05 seconds (v2.0.0-canary.12)
```

**Reply by rezaalavi** [reply]
Thanks, mate.

Is this for 11ty version 1 or 2?

 am interested in trying version 2 but I can't find any docs, all docs point to version 1, as a result, I have avoided V2.

At the moment I pass an environment variable to  --output and then read it back in my plugin but that won't make the plugin usable if I publish it on github. I also checked the code for the image-plugin but didn't find any methods there either and it seems we need to pass the output directory as an option to the plugin which might not be the best approach.

**Reply by pdehaan** [reply]
Seems to work in 1.0.1 as well:

```sh
npm run build

> 11ty-2483@1.0.0 build
> eleventy

{
  input: 'src',
  data: 'src/_data',
  includes: 'src/_includes',
  layouts: undefined,
  output: 'www'
}
[11ty] Writing www/index.html from ./src/index.njk
[11ty] Wrote 1 file in 0.03 seconds (v1.0.1)
```

```sh
npm run build2

> 11ty-2483@1.0.0 build2
> eleventy --output=bananasinpajamas

{
  input: 'src',
  data: 'src/_data',
  includes: 'src/_includes',
  layouts: undefined,
  output: 'bananasinpajamas'
}
[11ty] Writing bananasinpajamas/index.html from ./src/index.njk
[11ty] Wrote 1 file in 0.03 seconds (v1.0.1)
```

**Reply by KiwiKilian** [reply]
If I'm ready to commit this crime, awaiting the events: Is there any way to make eleventy wait for the initialization of my plugin? In my plugin I would first await the directories and then add a shortcode.

**Reply by zachleat** [reply]
> I found a dir in the eleventyConfig object from within my fake plugin. But that only seems to return whatever I set in my root .eleventy.js config file, and doesn't account for --output being set via the command line.

This is a bug, please file it!

**Reply by KiwiKilian** [reply]
Would be happy to do so, but I'm not sure which part is the bug. The `eleventyConfig` (instance of `UserConfig`) having the `dir` object or not representing `--output` set via the command line.

---

For my problem it seems to work fine to retrieve the `dir` from `eleventy.directories` [just like the `templateConfig` in the RenderPlugin](https://github.com/11ty/eleventy/blob/460caa2b3edd55c70402d492a6060b04708ccd6a/src/Plugins/RenderPlugin.js#L302-L305). Looks like this:

```js
module.exports = function (eleventyConfig, pluginConfig = {}) {
  let directoriesConfig;
  eleventyConfig.on("eleventy.directories", ( dir ) => {
    directoriesConfig = dir;
  });

  let templateConfig;
  eleventyConfig.on("eleventy.config", (config) => {
    templateConfig = config;
  });

  eleventyConfig.addAsyncShortcode("myShortcode", async () => {
    // Do stuff with directoriesConfig and templateConfig
  });
};

```

**Reply by zachleat** [reply]
In a plugin file I would expect this to work, as we run plugins in a second stage after the initial config callback function has completed:

```js
module.exports = function (eleventyConfig, pluginConfig = {}) {
  let directoriesConfig = eleventyConfig.dir;
};
```

If the above does not work, it is a bug!
If `--output` is not represented correctly in the `eleventyConfig.dir` I would consider that to also be a bug.

(Thank you!)

**Reply by pdehaan** [reply]
IIRC, the first one worked, but second one didn't.

```sh
npx @11ty/eleventy --input=pasta --output=noodles
[CONFIG] undefined
[PLUGIN] { input: 'src' }
[11ty] Writing noodles/index.html from ./pasta/index.njk
[11ty] Wrote 1 file in 0.03 seconds (v1.0.2)
```

Where my .eleventy.js is:
```js
const { inspect } = require("node:util");

/**
 * @param {import("@11ty/eleventy/src/UserConfig")} eleventyConfig
 * @returns {ReturnType<import("@11ty/eleventy/src/defaultConfig")>}
 */
module.exports = function (eleventyConfig) {
  console.log("[CONFIG]", inspect(eleventyConfig.dir, { sorted: true }));

  eleventyConfig.addPlugin(require("./plugins/custom"));

  return {
    dir: {
      input: "src",
      // output: "www",
    }
  };
};
```

And my ./plugins/custom.js is:
```js
const { inspect } = require("node:util");

module.exports = (eleventyConfig, pluginConfig) => {
  console.log("[PLUGIN]", inspect(eleventyConfig.dir, { sorted: true }));
  return {};
};
```

And my src/index.njk and pasta.njk are blank files that exist.

---

**UPDATE:** Pushed my sample repo to https://github.com/pdehaan/11ty-config-plugins.

```sh
npm run

Lifecycle scripts included in 11ty-config-plugins@1.0.0:
  test
    echo "Error: no test specified" && exit 1

available via `npm run-script`:
  build
    eleventy
  build:noodles
    eleventy --input=pasta --output=noodles
  build:pasta
    eleventy --input=./pasta
```

```sh
npm run build:noodles

> 11ty-config-plugins@1.0.0 build:noodles
> eleventy --input=pasta --output=noodles

[CONFIG] undefined
[PLUGIN] { input: 'src' }
[11ty] Writing noodles/index.html from ./pasta/index.njk
[11ty] Wrote 1 file in 0.03 seconds (v1.0.2)
```

I think it should resolve the `input` and `output` directories from the command line (and possibly return the defaults if they are not specified; like `output: "_site"`, etc).

**Reply by pdehaan** [reply]
@zachleat Filed as https://github.com/11ty/eleventy/issues/2729 for your bug tracking pleasure.

**Reply by zachleat** [reply]
Thank you @pdehaan !!
